src/libsystemd/sd-journal/journal-file.c

```diff
@@ -3549,15 +3549,14 @@ int journal_file_next_entry(
          * the same offset, and the index needs to be shifted. Otherwise, use the found object as is,
          * as it is the nearest entry object from the input offset 'p'. */
 
-        if (p != q)
-                goto found;
-
-        r = bump_array_index(&i, direction, n);
-        if (r <= 0)
-                return r;
+        if (p == q) {
+                r = bump_array_index(&i, direction, n);
+                if (r <= 0)
+                        return r;
+        }
 
         /* And jump to it */
-        r = generic_array_get(f, le64toh(f->header->entry_array_offset), i, direction, ret_object ? &o : NULL, &q);
+        r = generic_array_get(f, le64toh(f->header->entry_array_offset), i, direction, ret_object, &q);
```

Comment:
this collides with our usual coding style that we do not touch return params on failure. Let's say this call here succeeds and returns r > 0. Then the check_properly_ordered() check below might still fail, in which case we'll have touch ret_object, but return an error...

Reply:
Indeed. Good catch. The change is not essential here. I will fix it.